src/journal/journald-kmsg.c

```diff
@@ -263,12 +264,21 @@ void dev_kmsg_record(Server *s, char *p, size_t l) {
         if (LOG_FAC(priority) == LOG_KERN)
                 iovec[n++] = IOVEC_MAKE_STRING("SYSLOG_IDENTIFIER=kernel");
         else {
+                static bool safe_mode = false;
+
                 pl -= syslog_parse_identifier((const char**) &p, &identifier, &pid);
 
-                /* Avoid any messages we generated ourselves via
-                 * log_info() and friends. */
-                if (is_us(identifier, pid))
-                        goto finish;
+                /* Messages we generated ourselves via log_info() and log_debug() are fine until we
+                 * face some unexpected conditions. */
+                if (is_us(identifier, pid)) {
+                        if (LOG_PRI(priority) <= LOG_NOTICE)
+                                safe_mode = true;
+
+                        if (safe_mode)
+                                goto finish;
+
+                        iovec[n++] = IOVEC_MAKE_STRING("_SYSTEMD_UNIT=" SPECIAL_JOURNALD_SERVICE);
+                }
```

Comment:
I am sorry, but this is not ok. this will create log cycles, since we do some logging when opening journal files, and thus we might end up logging because we want to log, which is really not a good idea. This might overrun the stack frame.